web-app/HeimScripts/heatmap/run.R

```diff
@@ -1,187 +1,254 @@
+    length(label) - 2  #last too elements are node and subset.
+  label <- label[1:endOfSubject]
+  paste(label, collapse = "_")
 }
 
-buildFields <- function(df){
-  df <- melt(df, na.rm = T, id=c("Row.Label","Bio.marker","SIGNIFICANCE","MEAN","SD")) # melt implicitly casts characters to factors to make your like more exciting, in order to encourage more adventures it does not have characters.as.factors=F param.
-  ZSCORE <- (df$value - df$MEAN)/df$SD
+buildFields <- function(df) {
+  df <- melt(
+    df, na.rm = T, id = c("Row.Label",
+                          "Bio.marker",
+                          "SIGNIFICANCE",
+                          "MEAN",
+                          "SD")
+  )  # melt implicitly casts
+     # characters to factors to make your like more exciting,
+     # in order to encourage more adventures it does not
+     # have characters.as.factors=F param.
+  ZSCORE <- (df$value - df$MEAN) / df$SD
   df["MEAN"] <- NULL
-  df["SD"]   <-NULL
-  names(df) <- c("PROBE","GENESYMBOL","SIGNIFICANCE","PATIENTID","VALUE")
+  df["SD"]   <- NULL
+  names(df) <-
+    c("PROBE","GENESYMBOL","SIGNIFICANCE","PATIENTID","VALUE")
   df["ZSCORE"] <- ZSCORE
   return(df)
 }
 
-writeDataForZip <- function(df, patientIDs) {
-  pidCols    <- as.character(patientIDs)
-  t          <- df
-  t[pidCols] <- lapply(t[pidCols], function(v) { (v - df$MEAN) / df$SD })
-  allCols    <- c(c("Row.Label", "Bio.marker", "MEAN", "SD", "SIGNIFICANCE"), pidCols)
-  write.table(t[allCols], "heatmap_data.tsv", sep="\t", na="", row.names=F, col.names=T)
-
-  allCols <- c(c("Row.Label"), pidCols)
-  write.table(df[allCols], "heatmap_orig_values.tsv", sep="\t", na="", row.names=F, col.names=T)
+writeDataForZip <- function(df, zScores, patientIDs) {
+  pidCols <- as.character(patientIDs)
+  df      <- df[ , -which(names(df) %in% pidCols)]  # Drop patient columns
+  df      <- cbind(df,zScores)                      # Replace with zScores
+  write.table(
+    df,
+    "heatmap_data.tsv",
+    sep = "\t",
+    na = "",
+    row.names = FALSE,
+    col.names = TRUE
+  )
 }
 
 writeRunParams <- function(max_rows, sorting) {
-  params <- list(
-    max_rows = max_rows,
-    sorting  = sorting
-  )
-
+  params <- list(max_rows = max_rows,
+                 sorting  = sorting)
   if (exists("preprocessed") && exists("preprocessing_params")) {
     params <- c(params, preprocessing_params)
   }
-
-  write(toJSON(params, pretty=TRUE), 'params.json')
+  write(toJSON(params, pretty = TRUE), 'params.json')
 }
 
 fixString <- function(str) {
-  str <- gsub("[^a-zA-Z0-9-]", "", str, perl=TRUE)
-  return(str)
+  gsub("[^a-zA-Z0-9-]", "", str, perl = TRUE)
 }
 
-buildExtraFields <- function(df){
-  FEATURE <- rep("Cohort", nrow(df) )
+buildExtraFields <- function(df) {
+  FEATURE <- rep("Cohort", nrow(df))
   PATIENTID <- as.character(df$PATIENTID)
   TYPE <- rep("binary",nrow(df))
   VALUE <- getSubset(PATIENTID)
   extraFields <- data.frame(FEATURE, PATIENTID, TYPE, VALUE)
```

Comment:
Note: if done this way the columns will be of class factor. Add 'stringsAsFactors = F' or set default parameter of stringsAsFactors (try default.stringsAsFactors()) to FALSE


TYPE <- rep("binary",nrow(df))
VALUE <- letters[1:5]
 extraFields <- data.frame(TYPE, VALUE)
class(extraFields[,1])
[1] "factor"

extraFields <- data.frame(TYPE, VALUE, stringsAsFactors = F)
class(extraFields[,2])
[1] "character"